### PR TITLE
Lazily initializing exceptions

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
Delay the creation of `ApiCallTimeoutException` and `ApiCallAttemptTimeoutException` as they are not necessary and costly to create. This would improve performance in async path.

## Testing
Benchmark comparison (throughput ops/s)

Benchmark  |  before | after |     
------------ | -------------  | -----------                              
NettyClientH1NonTlsBenchmark.concurrentApiCall  |  7119 | 8453
NettyClientH1NonTlsBenchmark.sequentialApiCall |  2303 | 2393
NettyHttpClientH1Benchmark.concurrentApiCall (jdk) |  6390 | 7260
NettyHttpClientH1Benchmark.sequentialApiCall (jdk) |  1775 |  1827
NettyHttpClientH2Benchmark.concurrentApiCall (jdk) | 4513 | 4655
NettyHttpClientH2Benchmark.sequentialApiCall (jdk) | 1687 | 1723

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: build-tools/src/main/resources/software/amazon/awssdk/checkstyle-suppressions.xml [hunk 1]
```diff
@@ -42,4 +42,8 @@
     <!-- Assumes getter/setter match JSON property -->
     <suppress checks="AbbreviationAsWordInName"
               files=".*[\\/]software[\\/]amazon[\\/]awssdk[\\/]services[\\/]s3[\\/]event[\\/]S3EventNotification.java$"/>
+
+    <!-- This check is problematic with lambda. https://github.com/checkstyle/checkstyle/issues/3342 -->
+    <suppress checks="Indentation"
+              files="AsyncApiCallTimeoutTrackingStage.java|MakeAsyncHttpRequestStage.java"/>
```
Comment (millems) [inline build-tools/src/main/resources/software/amazon/awssdk/checkstyle-suppressions.xml]: Can we just work around it instead? Turning off indentation checking in general is not great, and we've worked around it elsewhere.

Comment (zoewangg) [reply build-tools/src/main/resources/software/amazon/awssdk/checkstyle-suppressions.xml]: I tried but I find it is very ugly to comply with the checkstyle rules when you have lambda as parameter.

```
        return TimerUtils.timeAsyncTaskIfNeeded(executeFuture,
                                                timeoutExecutor, 
           () -> ApiCallAttemptTimeoutException.create(timeoutMillis),
                                                timeoutMillis);
```

Comment (millems) [reply build-tools/src/main/resources/software/amazon/awssdk/checkstyle-suppressions.xml]: Can you factor it into a separate method, assign it to a variable, or wrap after the `->`?

Comment (zoewangg) [reply build-tools/src/main/resources/software/amazon/awssdk/checkstyle-suppressions.xml]: Yeah, creating variable would work but I was not sure if it was worth it to just comply with checkstyle rules. If we think keeping indention checkstyle is important enough, I can make the change.

**Comment by zoewangg** [comment]
> Why are they so expensive to create?

I think `Throwable.fillinstacktrace()` is very expensive and it is being invoked when an exception is being created.

File: build-tools/src/main/resources/software/amazon/awssdk/checkstyle-suppressions.xml (hunk 1 above)
Comment (millems) [reply build-tools/src/main/resources/software/amazon/awssdk/checkstyle-suppressions.xml]: If we could have a more targeted checkstyle rule, I'd favor disabling it, but the indentation checkstyle catches a lot of other issues.

Comment (zoewangg) [reply build-tools/src/main/resources/software/amazon/awssdk/checkstyle-suppressions.xml]: Yeah, updated

**Pull request merged**